com.unity.render-pipelines.high-definition/Runtime/Material/BuiltinGIUtilities.hlsl

```diff
@@ -157,12 +121,48 @@ float3 SampleBakedGI(PositionInputs posInputs, float3 normalWS, uint renderingLa
 
 #else // PROBEVOLUMESEVALUATIONMODES_MATERIAL_PASS || PROBEVOLUMESEVALUATIONMODES_DISABLED
 #if SAMPLE_PROBEVOLUME
+    if (_EnableProbeVolumes)
+    {
 #if SAMPLE_LIGHTMAP
-    float probeVolumeHierarchyWeight = 1.0f;
+        float probeVolumeHierarchyWeight = 1.0f;
 #else
-    float probeVolumeHierarchyWeight = 0.0f;
+        float probeVolumeHierarchyWeight = 0.0f;
+#endif
+
+#ifdef SHADERPASS
+#if SHADERPASS == SHADERPASS_GBUFFER || SHADERPASS == SHADERPASS_FORWARD
+#if SHADERPASS == SHADERPASS_GBUFFER || (SHADERPASS == SHADERPASS_FORWARD && defined(USE_FPTL_LIGHTLIST))
+        // posInputs.tileCoord will be zeroed out in GBuffer pass.
+        // posInputs.tileCoord will be incorrect for probe volumes (which use clustered) in forward if forward lightloop is using FTPL lightlist (i.e: in ForwardOnly lighting configuration). 
+        // Need to manually compute tile coord here.
+        float2 positionSS = posInputs.positionNDC.xy * _ScreenSize.xy;
+        uint2 tileCoord = uint2(positionSS) / ProbeVolumeGetTileSize();
+        posInputs.tileCoord = tileCoord;
+        #endif
+
+        // TODO: SampleBakedGI() will be called twice, once for the front facing direction, and once for the back facing direction (for transmission).
+        // For Probe Volumes specifically, this is not necessary - it would be better to return the coefficients, and just evaluate those coefficients twice (once for for each normal).
+        // We already do this in LightLoop evaluation mode.
```

Comment:
Either we use multi-compile or keepd it in shader config, in all cases we should hide this behind a single function:

combinedGI += EvaluateProbeVolumeSphericalHarmonics(posInputs, normalWS, renderingLayers, coefficients, probeVolumeHierarchyWeight);
        return  EvaluateProbeVolumeAmbientProbeFallback(normalWS, probeVolumeHierarchyWeight);

Which internally have 
if SHADEROPTIONS_PROBE_VOLUMES_ENCODING_MODE == PROBEVOLUMESENCODINGMODES_SPHERICAL_HARMONICS_L0
        ProbeVolumeSphericalHarmonicsL0 coefficients;
        ProbeVolumeAccumulateSphericalHarmonicsL0(posInputs, normalWS, renderingLayers, coefficients, probeVolumeHierarchyWeight);
        return EvaluateProbeVolumeSphericalHarmonicsL0(normalWS, coefficients);
#elif SHADEROPTIONS_PROBE_VOLUMES_ENCODING_MODE == PROBEVOLUMESENCODINGMODES_SPHERICAL_HARMONICS_L1
        ProbeVolumeSphericalHarmonicsL1 coefficients;
        ProbeVolumeAccumulateSphericalHarmonicsL1(posInputs, normalWS, renderingLayers, coefficients, probeVolumeHierarchyWeight);
        return = EvaluateProbeVolumeSphericalHarmonicsL1(normalWS, coefficients);
#elif SHADEROPTIONS_PROBE_VOLUMES_ENCODING_MODE == PROBEVOLUMESENCODINGMODES_SPHERICAL_HARMONICS_L2
        ProbeVolumeSphericalHarmonicsL2 coefficients;
        ProbeVolumeAccumulateSphericalHarmonicsL2(posInputs, normalWS, renderingLayers, coefficients, probeVolumeHierarchyWeight);
                return  EvaluateProbeVolumeSphericalHarmonicsL2(normalWS, coefficients);
#endif


This is not readable otherwise :)

Reply:
I agree, and want to move it into that direction. The one issue is that the generic version:
EvaluateProbeVolumeSphericalHarmonics() will need to be conditionally declared with ProbeVolumeSphericalHarmonicsL0, ProbeVolumeSphericalHarmonicsL1, or ProbeVolumeSphericalHarmonicsL2 coefficient types.

Or I could make the generic version only have args:
EvaluateProbeVolumeSphericalHarmonics(posInputs, normalWS, renderingLayers)
And it would also contain the call to EvaluateProbeVolumeAmbientProbeFallback() so that coefficients do not need to be returned in an out parameter.